Core/Configuration/JsonConfiguration.cs

```diff
@@ -0,0 +1,419 @@
+                {
+                    config.AutoStartInstance = value ?? "";
+
+                    SaveConfig();
+                }
+            }
+        }
+
+        // <summary>
+        // Create a new instance of Win32RegistryJson. ServiceLocator maintains a
+        // singleton instance, so in general you should use that. However, the
+        // core state is static, so creating multiple instances is not an issue.
+        // </summary>
+        public JsonConfiguration()
+        {
+            lock (_lock)
+            {
+                if (config != null)
+                    return;
+
+                LoadConfig();
+            }
+        }
+
+        // <summary>
+        // For testing purposes only. This constructor discards the global configuration
+        // state, and recreates it from the specified file.
+        // </summary>
+        //
+        // N.B., if you're adding the ability to specify a config file from the CLI, this
+        // might be the way to do it. However, you need to ensure that the configuration
+        // doesn't get loaded from the default location first, as that might end up
+        // creating files and directories that the user is trying to avoid creating by
+        // specifying the configuration file on the command line.
+        public JsonConfiguration (string newConfig)
+        {
+            lock (_lock)
+            {
+                configFile = newConfig;
+
+                LoadConfig();
+            }
+        }
+
+
+        public JBuilds GetKSPBuilds ()
+        {
+            lock (_lock)
+            {
+                return config.KSPBuilds;
+            }
+        }
+
+        public void SetKSPBuilds (JBuilds buildMap)
+        {
+            lock (_lock)
+            {
+                config.KSPBuilds = buildMap;
+
+                SaveConfig();
+            }
+        }
+
+        public IEnumerable<Tuple<string, string>> GetInstances ()
+        {
+            lock (_lock)
+            {
+                return config.KspInstances.Select(instance =>
+                    new Tuple<string, string>(instance.Name, instance.Path));
+            }
+        }
+
+        public void SetRegistryToInstances(SortedList<string, KSP> instances)
+        {
+            lock (_lock)
+            {
+                config.KspInstances = instances.Select(instance => new KspInstance
+                {
+                    Name = instance.Key,
+                    Path = instance.Value.GameDir()
+                }).ToList();
+
+                SaveConfig();
+            }
+        }
+
+        public IEnumerable<string> GetAuthTokenHosts()
+        {
+            lock (_lock)
+            {
+                return config.AuthTokens.Select(token => token.Host);
+            }
+        }
+
+
+        public bool TryGetAuthToken (string host, out string token)
+        {
+            lock (_lock)
+            {
+                foreach (AuthToken t in config.AuthTokens)
```

Comment:
This block would be simpler if `AuthTokens` was a `Dictionary<string, string>`. We would just call `config.AuthTokens.TryGetValue(host, out token)` and handle the missing case in an `else`.

Reply:
That's a good point. The current design was chosen to make the config file clearer with host and token explicitly labelled, but I don't think much clarity would be lost by switching to a dictionary.